lib/identity/index.js

```diff
@@ -0,0 +1,185 @@
+ * License for the specific language governing rights and limitations under
+ * the License. Portions of the software are Copyright (c) TOZNY LLC, 2018-19.
+ * All rights reserved.
+ *
+ * @copyright Copyright (c) 2018-19 Tozny, LLC (https://tozny.com)
+ */
+
+'use strict'
+
+import Client from './client'
+import Config from './config'
+import 'isomorphic-fetch'
+import { checkStatus } from '../utils'
+import CryptoConsumer from '../utils/cryptoConsumer'
+import { Config as StorageConfig } from '../storage'
+import { PublicKey, SigningKey } from '../types'
+
+async function deriveNoteCreds(id, username, password) {
+  const noteID = await id.crypto.genericHash(username + id.config.realmId)
+  const cryptoKeyPair = await id.crypto.deriveCryptoKey(
+    password,
+    id.config.realmId + id.config.realmName,
+    10000
+  )
+  const signingKeyPair = await id.crypto.deriveSigningKey(
+    password,
+    cryptoKeyPair.publicKey + cryptoKeyPair.privateKey,
+    10000
+  )
+  return { noteID, cryptoKeyPair, signingKeyPair }
+}
+
+function createClient(id, userId, storageConfig) {
+  const idConfig = id.config.clone({ userId: userId })
+  const storageClient = new id.StorageClient(storageConfig)
+  return new Client(idConfig, storageClient, id.crypto)
+}
+
+export default class Identity extends CryptoConsumer {
+  static get Client() {
+    return Client
+  }
+
+  static get Config() {
+    return Config
+  }
+
+  /**
+   * Abstract getter for a storage Client constructor function.
+   *
+   * When implementing this class, this getter must be overloaded. When called it
+   * should offer up a storage Client constructor function. Identity constructs
+   * storage clients as part of creating Identity clients.
+   *
+   * An additional instance level getter is also provided which allows fetching
+   * the storage client constructor in both static _and_ instance method
+   * contexts as `this.StorageClient`.
+   *
+   * @returns {Client} The storage Client constructor.
+   */
+  static get StorageClient() {
+    throw new Error(
+      'Implementing classes must overloaded the StorageClient method to provide a valid storage Client constructor.'
+    )
+  }
+
+  constructor(config) {
+    super()
+    this.config = config
+  }
+
+  /**
+   * Allows `this.StorageClient` syntax in instance methods.
+   *
+   * Gets the static StorageClient constructor available in the static class. By
+   * returning it as a getter `this.StorageClient` syntax is support in
+   * instance methods.
+   *
+   * @returns {Client} The storage Client constructor.
+   */
+  get StorageClient() {
+    // Use this.constructor to ensure we referencing the implementing class, not an interface class.
+    return this.constructor.StorageClient
+  }
+
+  async register(username, password, token) {
+    const cryptoKeys = await this.StorageClient.generateKeypair()
+    const signingKeys = await this.StorageClient.generateSigningKeypair()
```

Comment:
nit: these key generation methods should live on the top level Tozny/helper object versus the StorageClient.

Reply:
They need access to the crypto object. We could put together a smarter helper object. `helpers` was intended mostly to be platform specific helpers, such as a way of pulling a profile from the file system ~/.tozny folder. Makes sens in node, but not at all in the browser context. At the very lease these should stay in the storage Client for backwards compatibility reasons. I'm thinking we should more directly expose the crypto implementation on the top level Tozny object? Will figure out something :)

Reply:
This interface does not directly determine the shape of an implementation's export. I'm wondering if it somehow should. Perhaps a default method in index.js which takes all the pieces and exports the 'Tozny' object. This helps ensure the default package exports keep the same shape. Still not sure the best shape for that, yet, though. I'm going to leave it looser with an eye towards tightening it as we get to a stable release state.

In the mean time, I have exposed the crypto implementation in the browser SDK directly from the Tozny object:

```
keyPair = await Tozny.crypto.generateKeyPair() / o/r any other defined crypto operation available there
```